servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/RedirectingClientAndConnectionFilterTest.java

```diff
@@ -15,209 +15,176 @@
-            assertThat(response.status(), equalTo(OK));
-        }
+
+        BlockingHttpRequester client = asBlockingRequester(createFilter((responseFactory, request) -> {
+            if (request.requestTarget().equals("/")) {
+                return Single.succeeded(responseFactory.permanentRedirect()
+                        .addHeader(LOCATION, "/next"));
+            }
+            return Single.succeeded(responseFactory.ok());
+        }, new RedirectingHttpRequesterFilterFactory(new RedirectingHttpRequesterFilter(), remoteAddress())));
+        HttpRequest request = client.get("/").addHeader(HOST, "servicetalk.io");
+        HttpResponse response = client.request(noOffloadsStrategy(), request);
+        assertThat(response.status(), equalTo(PERMANENT_REDIRECT));
+
+        response = client.request(noOffloadsStrategy(), request.addHeader("X-REDIRECT", "TRUE"));
+        assertThat(response.status(), equalTo(OK));
     }
 
     @Test
     public void redirectFilterWithHostHeaderAbsoluteLocation() throws Exception {
-        try (ServerContext serverContext = HttpServers.forAddress(localAddress(0))
-                .listenBlockingAndAwait((ctx, request, responseFactory) -> {
-                    if (request.requestTarget().equals("/")) {
-                        return responseFactory.permanentRedirect()
-                                .addHeader(LOCATION, "http://servicetalk.io/next");
-                    }
-                    return responseFactory.ok();
-                }); BlockingHttpRequester client = newRequester(serverContext)) {
-
-            HttpRequest request = client.get("/").addHeader(HOST, "servicetalk.io");
-            HttpResponse response = client.request(noOffloadsStrategy(), request);
-            assertThat(response.status(), equalTo(PERMANENT_REDIRECT));
-
-            response = client.request(noOffloadsStrategy(), request.addHeader("X-REDIRECT", "TRUE"));
-            assertThat(response.status(), equalTo(OK));
+
+        BlockingHttpRequester client = asBlockingRequester(createFilter((responseFactory, request) -> {
+            if (request.requestTarget().equals("/")) {
+                return Single.succeeded(responseFactory.permanentRedirect()
+                        .addHeader(LOCATION, "http://servicetalk.io/next"));
+            }
+            return Single.succeeded(responseFactory.ok());
+        }, new RedirectingHttpRequesterFilterFactory(new RedirectingHttpRequesterFilter(), remoteAddress())));
+        HttpRequest request = client.get("/").addHeader(HOST, "servicetalk.io");
+        HttpResponse response = client.request(noOffloadsStrategy(), request);
+        assertThat(response.status(), equalTo(PERMANENT_REDIRECT));
+
+        response = client.request(noOffloadsStrategy(), request.addHeader("X-REDIRECT", "TRUE"));
+        assertThat(response.status(), equalTo(OK));
+    }
+
+    private static class RedirectingHttpRequesterFilterFactory
```

Comment:
Can you explain the motivation for this filter factory?

Previously, the test wasn't using a host header filter or doing the strategy aware influencer conversions. Can you also explain the reasons for adding those?

Reply:
Yes, sure! 

1) the strategy aware influencer conversions
In previous version of RedirectingClientAndConnectionFilterTest [1] redirect filter is added using `HttpClientBuilder.appendClientFilter` that creates underneath the `StrategyInfluencerAwareConversions.toConditionalClientFilterFactory` filter [2]

2) host header filter
In previous version of RedirectingClientAndConnectionFilterTest uses `HttpClients.forSingleAddress` -> `DefaultSingleAddressHttpClientBuilder.forHostAndPort` which adds host header filter [5]

If we remove host header filter from `RedirectingHttpRequesterFilterFactory` we will get null here `redirectHost = request.effectiveHost();` [6]

[1] https://github.com/apple/servicetalk/blob/56e0dd753060fd4f0b7e43ab9bf4e77906b608e1/servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/RedirectingClientAndConnectionFilterTest.java#L91

[2] https://github.com/apple/servicetalk/blob/56e0dd753060fd4f0b7e43ab9bf4e77906b608e1/servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpClientBuilder.java#L142

[3] https://github.com/apple/servicetalk/blob/56e0dd753060fd4f0b7e43ab9bf4e77906b608e1/servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/RedirectingClientAndConnectionFilterTest.java#L89

[4] https://github.com/apple/servicetalk/blob/56e0dd753060fd4f0b7e43ab9bf4e77906b608e1/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpClients.java#L128

[5] https://github.com/apple/servicetalk/blob/56e0dd753060fd4f0b7e43ab9bf4e77906b608e1/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java#L289

[6] https://github.com/apple/servicetalk/blob/56e0dd753060fd4f0b7e43ab9bf4e77906b608e1/servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java#L267

Reply:
Ok now I understand the limitations with the task of using `AbstractHttpRequesterFilterTest` as it does not use the client builder and the conditional filter is package private. I apologize that this task was not as straight forward as expected for a "good first issue" and hence the back and forth.

The changes here are not ideal as the `HttpExecutionStrategyInfluencer` usage is not required for this test (the influencer is used only when there is an intent to "influence" the offloading[1]). Using the influencer related functionality makes this a bit confusing. I would suggest instead create a generic conditional filter factory class that can be utilized here to conditionally add the redirect filter and also append the host header filter while preserving the filter factory contract (implementing both `StreamingHttpClientFilterFactory` and `StreamingHttpConnectionFilterFactory`). Such a filter factory may look like:

```java
public final class ConditionalFilterFactory
        implements StreamingHttpConnectionFilterFactory, StreamingHttpClientFilterFactory{
    private final Predicate<StreamingHttpRequest> predicate;
    private final FilterFactory predicateFactory;

    public ConditionalFilterFactory(final Predicate<StreamingHttpRequest> predicate,
                                    final FilterFactory predicateFactory) {
        this.predicate = predicate;
        this.predicateFactory = predicateFactory;
    }

    @Override
    public StreamingHttpClientFilter create(final FilterableStreamingHttpClient client) {
        return new ConditionalHttpClientFilter(predicate, predicateFactory.create(client), client);
    }

    @Override
    public StreamingHttpConnectionFilter create(final FilterableStreamingHttpConnection connection) {
        return new ConditionalHttpConnectionFilter(predicate, predicateFactory.create(connection), connection);
    }

    public FilterFactory append(FilterFactory append) {
        StreamingHttpClientFilterFactory clientFactory = append((StreamingHttpClientFilterFactory) append);
        StreamingHttpConnectionFilterFactory connectionFactory = append((StreamingHttpConnectionFilterFactory) append);
        return new FilterFactory() {
            @Override
            public StreamingHttpClientFilter create(final FilterableStreamingHttpClient client) {
                return clientFactory.create(client);
            }

            @Override
            public StreamingHttpConnectionFilter create(final FilterableStreamingHttpConnection connection) {
                return connectionFactory.create(connection);
            }
        };
    }

    public interface FilterFactory extends StreamingHttpClientFilterFactory, StreamingHttpConnectionFilterFactory {

        static <FF extends StreamingHttpClientFilterFactory & StreamingHttpConnectionFilterFactory> FilterFactory from(
                FF original) {
            return new FilterFactory() {
                @Override
                public StreamingHttpClientFilter create(final FilterableStreamingHttpClient client) {
                    return original.create(client);
                }

                @Override
                public StreamingHttpConnectionFilter create(final FilterableStreamingHttpConnection connection) {
                    return original.create(connection);
                }
            };
        }
    }
}
```

This factory can be used in the redirect test as follows:

```java
private FilterFactory newFilterFactory() {
    return new ConditionalFilterFactory(request -> request.headers().contains("X-REDIRECT"),
            FilterFactory.from(new RedirectingHttpRequesterFilter()))
            .append(FilterFactory.from(
                    new HostHeaderHttpRequesterFilter(HostAndPort.of(remoteAddress()).toString())));
}
```
[1] https://apple.github.io/servicetalk/servicetalk-http-api/SNAPSHOT/blocking-safe-by-default.html#user-code

Reply:
Thank you Nitesh for detailed explanation!